app/models/topic_tracking_state.rb

```diff
@@ -293,12 +321,22 @@ def self.report(user, topic_id = nil)
       SQL
     end
 
-    DB.query(
+    report = DB.query(
       sql,
         user_id: user.id,
         topic_id: topic_id,
         min_new_topic_date: Time.at(SiteSetting.min_new_topics_time).to_datetime
     )
+
+    (@refine_methods || []).each do |refinement|
+      report = refinement.call(report, user)
+    end
+
+    report = report.uniq do |tracking_state|
```

Comment:
not happy with uniq being called unconditionally. In fact maybe part of the contract shall be ... refiner shall not introduce dupes ... then you don't need this at all.

Reply:
Hmm in that case I probably need to pass some data structure class to the refine method instead of the raw array, then that can control this e.g.

```ruby
class TopicTrackingStateReport do
  def initialize(report_data)
    @report_data = report_data
    @report_data_map = {}
    @report_data.each do |data|
      if !@report_data_map.key?(data[:topic_id])
        @report_data_map[data[:topic_id]] = data
      end
    end
  end

  def add(data)
    if @report_data_map.key?(data.topic_id) raise NotUniqueError
    @report_data_map[data.topic_id] = data
  end

  def read
    @report_data
  end
end
```

I think I should turn the data array into a hash to speed up the lookups. This way `refinement.call` would pass through an instance of this class. Thoughts? I don't think we want to trust the refiner methods to not introduce dupes